# Case study - birds from the UK

In this example, we will build a map of the observations of birds from the
United Kingdom.

In [1]:
using GBIF

## Getting the data

The first thing is to specify the `taxonKey` -- this is explained in the [GBIF
taxonomy][tax] (but lookup feature will be added to the package later), and the
country code (`GB`). Because the GBIF API cannot return more than 200000 records
(that's already quite a lot, but common species have been observed far more), we
will (i) focus on a single common species (the House sparrow, whose `taxonKey` is `5231190`), and
only observations for the year 2015.

[tax]: http://www.gbif.org/species/search?dataset_key=d7dddbf4-2cf0-4f39-9b2a-bb099caae36c

In [2]:
uk_birds_query = Dict(
  "taxonKey"=>5231190,
  "country"=>"GB",
  "hasCoordinate"=>true,
  "year"=>2015)
uk_birds = occurrences(uk_birds_query)

A  list of occurrences with a custom query - 20 out of 6296


This is good! The output means that we have 20 records already, out of 6296 that exist in GBIF. By default, we will only get the first 20, because (i) it avoids downloading all of the records at once, which consumes GBIF resources and (ii) there is a very simple way to get the records later.

We can now complete the request. This would take a while, since we are
going through occurrences 20 at a time (this is the GBIF default). We can speed up the process a little
bit, by requesting 200 records on every page. GBIF will not let you ask for more
than 200 records by page.

In [3]:
uk_birds.query["limit"] = 200

200

Let's now complete the request for real:

In [4]:
complete!(uk_birds)

Let's just have a look at the object now:

In [5]:
uk_birds

A  list of occurrences with a custom query - 6296 out of 6296


Once this is done, we can start looking at observations. Let's load the `Plots`
package to help us. We will also use `DataFrames`, because `GBIF` can export the results directly.

In [6]:
using Plots, DataFrames, StatPlots
#pyplot() # Feel free to replace by your favorite backend

In [7]:
birbs = DataFrame(uk_birds)

6296×36 DataFrames.DataFrame
│ Row  │ key        │ datasetKey                             │
├──────┼────────────┼────────────────────────────────────────┤
│ 1    │ 1344121861 │ "4fa7b334-ce0d-4e88-aaae-2e0c138d049e" │
│ 2    │ 1344129843 │ "4fa7b334-ce0d-4e88-aaae-2e0c138d049e" │
│ 3    │ 1347076448 │ "4fa7b334-ce0d-4e88-aaae-2e0c138d049e" │
│ 4    │ 1347246579 │ "4fa7b334-ce0d-4e88-aaae-2e0c138d049e" │
│ 5    │ 1347291334 │ "4fa7b334-ce0d-4e88-aaae-2e0c138d049e" │
│ 6    │ 1348757961 │ "4fa7b334-ce0d-4e88-aaae-2e0c138d049e" │
│ 7    │ 1344852419 │ "4fa7b334-ce0d-4e88-aaae-2e0c138d049e" │
│ 8    │ 1344964843 │ "4fa7b334-ce0d-4e88-aaae-2e0c138d049e" │
│ 9    │ 1345331295 │ "4fa7b334-ce0d-4e88-aaae-2e0c138d049e" │
│ 10   │ 1346245257 │ "4fa7b334-ce0d-4e88-aaae-2e0c138d049e" │
│ 11   │ 1411319383 │ "4fa7b334-ce0d-4e88-aaae-2e0c138d049e" │
⋮
│ 6285 │ 1400155848 │ "4fa7b334-ce0d-4e88-aaae-2e0c138d049e" │
│ 6286 │ 1401175477 │ "4fa7b334-ce0d-4e88-aaae-2e0c138d049e" │
│ 6287 │ 1401262703 │ "4fa7b334-ce0d-4e88-aaae-2e0c138d049e" │
│ 6288 │ 1401803366 │ "4fa7b334-ce0d-4e88-aaae-2e0c138d049e" │
│ 6289 │ 1402173043 │ "4fa7b334-ce0d-4e88-aaae-2e0c138d049e" │
│ 6290 │ 1403390242 │ "4fa7b334-ce0d-4e88-aaae-2e0c138d049e" │
│ 6291 │ 1403812398 │ "4fa7b334-ce0d-4e88-aaae-2e0c138d049e" │
│ 6292 │ 1404085653 │ "4fa7b334-ce0d-4e88-aaae-2e0c138d049e" │
│ 6293 │ 1550202327 │ "3a0e9440-0a97-4a04-8089-7398c77da0be" │
│ 6294 │ 1571836817 │ "0bbbf728-2d22-46fb-b5ab-37c353fc99ff" │
│ 6295 │ 1550611106 │ "3a0e9440-0a97-4a04-8089-7398c77da0be" │
│ 6296 │ 1553148181 │ "523590f0-f87d-4fed-acd2-6873bdce0a8c" │

│ Row  │ publishingOrgKey                       │ publishingCountry │
├──────┼────────────────────────────────────────┼───────────────────┤
│ 1    │ "e2e717bf-551a-4917-bdc9-4fa0f342c530" │ "US"              │
│ 2    │ "e2e717bf-551a-4917-bdc9-4fa0f342c530" │ "US"              │
│ 3    │ "e2e717bf-551a-4917-bdc9-4fa0f342c530" │ "US"              │
│ 4    │ "e2e717bf-551a-4917-bdc9-4fa0f342c530" │ "US"              │
│ 5    │ "e2e717bf-551a-4917-bdc9-4fa0f342c530" │ "US"              │
│ 6    │ "e2e717bf-551a-4917-bdc9-4fa0f342c530" │ "US"              │
│ 7    │ "e2e717bf-551a-4917-bdc9-4fa0f342c530" │ "US"              │
│ 8    │ "e2e717bf-551a-4917-bdc9-4fa0f342c530" │ "US"              │
│ 9    │ "e2e717bf-551a-4917-bdc9-4fa0f342c530" │ "US"              │
│ 10   │ "e2e717bf-551a-4917-bdc9-4fa0f342c530" │ "US"              │
│ 11   │ "e2e717bf-551a-4917-bdc9-4fa0f342c530" │ "US"              │
⋮
│ 6285 │ "e2e717bf-551a-4917-bdc9-4fa0f342c530" │ "US"              │
│ 6286 │ "e2e717bf-551a-4917-bdc9-4fa0f342c530" │ "US"              │
│ 6287 │ "e2e717bf-551a-4917-bdc9-4fa0f342c530" │ "US"              │
│ 6288 │ "e2e717bf-551a-4917-bdc9-4fa0f342c530" │ "US"              │
│ 6289 │ "e2e717bf-551a-4917-bdc9-4fa0f342c530" │ "US"              │
│ 6290 │ "e2e717bf-551a-4917-bdc9-4fa0f342c530" │ "US"              │
│ 6291 │ "e2e717bf-551a-4917-bdc9-4fa0f342c530" │ "US"              │
│ 6292 │ "e2e717bf-551a-4917-bdc9-4fa0f342c530" │ "US"              │
│ 6293 │ "48f02f0c-5287-4b0a-bdf1-cee8f9665285" │ "GB"              │
│ 6294 │ "db91efbd-8f97-42ec-8118-270dfab1d41b" │ "GB"              │
│ 6295 │ "48f02f0c-5287-4b0a-bdf1-cee8f9665285" │ "GB"              │
│ 6296 │ "25318ae9-d1de-456f-baa2-4515ff1f33b1" │ "GB"              │

│ Row  │ countryCode │ country          │ basisOfRecord     │ individualCount │
├──────┼─────────────┼──────────────────┼───────────────────┼─────────────────┤
│ 1    │ "GB"        │ "United Kingdom" │ HUMAN_OBSERVATION │ NA              │
│ 2    │ "GB"        │ "United Kingdom" │ HUMAN_OBSERVATION │ NA              │
│ 3    │ "GB"        │ "United Kingdom" │ HUMAN_OBSERVATION │ 30              │
│ 4    │ "GB"        │ "United Kingdom" │ HUMAN_OBSERVATION │ 12              │
│ 5    │ "GB"        │ "United Kingdom" │ HUMAN_OBSERVATION │ 2               │
│ 6    │ "GB"        │ "United Kingdom" │ HUMAN_OBSERVATION │ 2               │


Let's add a column for the month:

In [17]:
birbs[:month] = Dates.month.(birbs[:date]);

ERROR (unhandled task failure): InterruptException:
Stacktrace:
 [1] (::Base.##300#301{IJulia.#send_stderr,Timer})() at ./event.jl:430


In [9]:
histogram(birbs, :month,
  xticks=([1,3,5,7,9,11],["Jan","Mar","May","Jul","Sep","Nov"]),
  c=:lightgrey, lab="", ylab="Number of occurrences")

<?xml version="1.0" encoding="utf-8"?>
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 Jan 
 
 
 Mar 
 
 
 May 
 
 
 Jul 
 
 
 Sep 
 
 
 Nov 
 
 
 0 
 
 
 200 
 
 
 400 
 
 
 600 
 
 
 month 
 
 
 Number of occurrences

We can now have a look at the latitude / longitude of the points:

In [16]:
scatter(birbs, :longitude, :latitude, aspectratio=1.5, marker=(:circle, :grey, 0.5), leg=false)

<?xml version="1.0" encoding="utf-8"?>
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 -6 
 
 
 -4 
 
 
 -2 
 
 
 0 
 
 
 2 
 
 
 50 
 
 
 55 
 
 
 60 
 
 
 longitude 
 
 
 latitude 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
<circle clip-path="url(#clip02)" style="fill:#000000; stro

And voilà!